File: rust/crates/corlinman-canvas/src/adapters/table.rs
//! `table` artifact adapter — GFM markdown or CSV → `<table>`.
//!
//! Producers send exactly one of `markdown` or `csv` in
//! [`ArtifactBody::Table`]; this adapter rejects the both-present
//! and neither-present cases with [`CanvasError::Adapter`]. The
//! protocol-level shape is enforced by serde; this is a semantic
//! check.
//!
//! ## Markdown path — `pulldown-cmark`
//!
//! GFM tables are parsed with the `Tables` extension on. We walk
//! the event stream ourselves rather than calling `push_html`
//! because:
//!
//! - `push_html` wraps cells in `<p>` tags. We want compact
//!   `<td>cell</td>`.
//! - We need to stamp `cn-canvas-table` / `cn-canvas-table-row` /
//!   `cn-canvas-table-cell` classes for Tidepool styling. The
//!   built-in writer doesn't expose hook points.
//!
//! Inline content (bold, code spans) inside a cell is rendered as
//! escaped text in this iter — full inline support is iter 4
//! (latex) territory and shares no code with the table walker.
//!
//! ## CSV path — `csv` crate
//!
//! Standard reader, headers from the first record. Quote handling
//! and embedded commas are the crate's job. We HTML-escape every
//! cell before emission.
//!
//! ## Output shape
//!
//! ```html
//! <table class="cn-canvas-table">
//!   <thead><tr class="cn-canvas-table-row">
//!     <th class="cn-canvas-table-cell">a</th> ...
//!   </tr></thead>
//!   <tbody>
//!     <tr class="cn-canvas-table-row">
//!       <td class="cn-canvas-table-cell">1</td> ...
//!     </tr>
//!   </tbody>
//! </table>
//! ```

use pulldown_cmark::{Event, Options, Parser, Tag, TagEnd};

use crate::protocol::{ArtifactKind, CanvasError, RenderedArtifact, ThemeClass};

/// Outer wrapper class. Pairs with `border-tp-glass-edge
/// bg-tp-glass-inner` rules in iter 9's globals.css.
const WRAPPER_CLASS: &str = "cn-canvas-table";
const ROW_CLASS: &str = "cn-canvas-table-row";
const CELL_CLASS: &str = "cn-canvas-table-cell";

/// Adapter entry. Exactly one of `markdown` / `csv` must be present.
pub fn render(
    markdown: Option<&str>,
    csv_src: Option<&str>,
    theme_class: ThemeClass,
) -> Result<RenderedArtifact, CanvasError> {
    let html = match (markdown, csv_src) {
        (Some(md), None) => render_markdown(md)?,
        (None, Some(csv)) => render_csv(csv)?,
        (Some(_), Some(_)) => {
            return Err(CanvasError::Adapter {
                kind: ArtifactKind::Table,
                message: "table body must specify exactly one of `markdown` or `csv`, not both"
                    .to_string(),
            });
        }
        (None, None) => {
            return Err(CanvasError::Adapter {
                kind: ArtifactKind::Table,
                message: "table body must specify either `markdown` or `csv`".to_string(),
            });
        }
    };

    Ok(RenderedArtifact {
        html_fragment: html,
        theme_class,
        content_hash: String::new(), // iter 7 (cache) populates
        render_kind: ArtifactKind::Table,
        warnings: Vec::new(),
    })
}

/// Walk the pulldown-cmark event stream and emit a single
/// `<table>` element. Non-table events (paragraphs around the
/// table, etc.) are ignored — producers should send only a table.
/// If no table event is seen at all, return an adapter error so
/// the producer learns about it.
fn render_markdown(source: &str) -> Result<String, CanvasError> {
    let mut opts = Options::empty();
    opts.insert(Options::ENABLE_TABLES);
    let parser = Parser::new_ext(source, opts);

    let mut html = String::new();
    let mut in_thead = false;
    let mut saw_table = false;

    for event in parser {
        match event {
            Event::Start(Tag::Table(_)) => {
                saw_table = true;
                html.push_str("<table class=\"");
                html.push_str(WRAPPER_CLASS);
                html.push_str("\">");
            }
            Event::End(TagEnd::Table) => {
                html.push_str("</table>");
            }
            Event::Start(Tag::TableHead) => {
                in_thead = true;
                html.push_str("<thead><tr class=\"");
                html.push_str(ROW_CLASS);
                html.push_str("\">");
            }
            Event::End(TagEnd::TableHead) => {
                in_thead = false;
                html.push_str("</tr></thead><tbody>");
            }
            Event::Start(Tag::TableRow) => {
                html.push_str("<tr class=\"");
                html.push_str(ROW_CLASS);
                html.push_str("\">");
            }
            Event::End(TagEnd::TableRow) => {
                html.push_str("</tr>");
            }
            Event::Start(Tag::TableCell) => {
                html.push_str(if in_thead {
                    "<th class=\""
                } else {
                    "<td class=\""
                });
                html.push_str(CELL_CLASS);
                html.push_str("\">");
            }
            Event::End(TagEnd::TableCell) => {
                html.push_str(if in_thead { "</th>" } else { "</td>" });
            }
            Event::Text(t) => {
                push_escaped(&mut html, &t);
            }
            Event::Code(c) => {
                // Inline `code` spans inside a cell — render as a
                // small `<code>` so they're at least
                // distinguishable; the cell class still applies.
                html.push_str("<code>");
                push_escaped(&mut html, &c);
                html.push_str("</code>");
            }
            Event::Html(raw) | Event::InlineHtml(raw) => {
                // pulldown-cmark passes through raw HTML as `Html`
                // events. We *never* emit raw HTML inside a cell —
                // a `<script>` in a producer's table source must
                // surface as escaped text, not executable markup.
                push_escaped(&mut html, &raw);
            }
            // Drop everything else (SoftBreak, HardBreak,
            // FootnoteReference, …) — out of scope for table
            // cells. iter 4 (latex) revisits inline rendering.
            _ => {}
        }
    }

    // pulldown-cmark closes `<tbody>` implicitly via the parser
    // state machine; we matched only on Tag, so we have to close
    // ourselves.
    if let Some(idx) = html.rfind("</table>") {
        // Insert `</tbody>` before the closing `</table>` if not
        // already there. Cheap, correct.
        if !html[..idx].ends_with("</tbody>") {
            html.replace_range(idx..idx, "</tbody>");
        }
    }

    if !saw_table {
        return Err(CanvasError::Adapter {
            kind: ArtifactKind::Table,
            message: "markdown source contained no GFM table".to_string(),
        });
    }
    Ok(html)
}

/// CSV → `<table>`. First record is the header.
fn render_csv(source: &str) -> Result<String, CanvasError> {
    let mut reader = csv::ReaderBuilder::new()
        .has_headers(true)
        .flexible(true)
        .from_reader(source.as_bytes());

    let headers = reader
        .headers()
        .map_err(|e| CanvasError::Adapter {
            kind: ArtifactKind::Table,
            message: format!("CSV header parse failed: {e}"),
        })?
        .clone();

    let mut html = String::new();
    html.push_str("<table class=\"");
    html.push_str(WRAPPER_CLASS);
    html.push_str("\">");

    if !headers.is_empty() {
        html.push_str("<thead><tr class=\"");
        html.push_str(ROW_CLASS);
        html.push_str("\">");
        for cell in &headers {
            html.push_str("<th class=\"");
            html.push_str(CELL_CLASS);
            html.push_str("\">");
            push_escaped(&mut html, cell);
            html.push_str("</th>");
        }
        html.push_str("</tr></thead>");
    }

    html.push_str("<tbody>");
    for record in reader.records() {
        let record = record.map_err(|e| CanvasError::Adapter {
            kind: ArtifactKind::Table,
            message: format!("CSV row parse failed: {e}"),
        })?;
        html.push_str("<tr class=\"");
        html.push_str(ROW_CLASS);
        html.push_str("\">");
        for cell in &record {
            html.push_str("<td class=\"");
            html.push_str(CELL_CLASS);
            html.push_str("\">");
            push_escaped(&mut html, cell);
            html.push_str("</td>");
        }
        html.push_str("</tr>");
    }
    html.push_str("</tbody></table>");

    Ok(html)
}

/// HTML escape into an existing buffer to avoid the temporary
/// allocation `format!` would do. Five-char OWASP set.
fn push_escaped(out: &mut String, input: &str) {
    for ch in input.chars() {
        match ch {
            '&' => out.push_str("&amp;"),
            '<' => out.push_str("&lt;"),
            '>' => out.push_str("&gt;"),
            '"' => out.push_str("&quot;"),
            '\'' => out.push_str("&#39;"),
            other => out.push(other),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn push_escaped_handles_owasp_five() {
        let mut out = String::new();
        push_escaped(&mut out, "<a href=\"x\">'&'</a>");
        assert!(!out.contains('<'));
        assert!(!out.contains('>'));
        assert!(out.contains("&lt;"));
        assert!(out.contains("&quot;"));
        assert!(out.contains("&#39;"));
        assert!(out.contains("&amp;"));
    }
}

File: python/packages/corlinman-canvas/src/corlinman_canvas/table.py
"""`table` artifact adapter — GFM markdown or CSV → ``<table>``.

Python port of Rust ``adapters/table.rs``. Markdown is parsed with
``markdown-it-py`` (GFM tables extension); CSV uses stdlib ``csv``. HTML
emission is hand-rolled so we can stamp the Tidepool class names and
keep cells compact (``<td>cell</td>``, no wrapper ``<p>``).

Output shape (byte-equivalent to the Rust crate)::

    <table class="cn-canvas-table">
      <thead><tr class="cn-canvas-table-row">
        <th class="cn-canvas-table-cell">a</th> ...
      </tr></thead>
      <tbody>
        <tr class="cn-canvas-table-row">
          <td class="cn-canvas-table-cell">1</td> ...
        </tr>
      </tbody>
    </table>
"""

from __future__ import annotations

import csv as _csv
import io

from markdown_it import MarkdownIt

from .protocol import (
    AdapterError,
    ArtifactKind,
    RenderedArtifact,
    ThemeClass,
)

WRAPPER_CLASS = "cn-canvas-table"
ROW_CLASS = "cn-canvas-table-row"
CELL_CLASS = "cn-canvas-table-cell"


def _push_escaped(out: list[str], src: str) -> None:
    """Append HTML-escaped src into out; five-char OWASP set."""
    for ch in src:
        if ch == "&":
            out.append("&amp;")
        elif ch == "<":
            out.append("&lt;")
        elif ch == ">":
            out.append("&gt;")
        elif ch == '"':
            out.append("&quot;")
        elif ch == "'":
            out.append("&#39;")
        else:
            out.append(ch)


def render(
    markdown: str | None,
    csv_src: str | None,
    theme_class: ThemeClass,
) -> RenderedArtifact:
    """Render a ``table`` artifact. Exactly one of ``markdown`` /
    ``csv_src`` must be set."""

    if markdown is not None and csv_src is not None:
        raise AdapterError(
            ArtifactKind.TABLE,
            "table body must specify exactly one of `markdown` or `csv`, not both",
        )
    if markdown is None and csv_src is None:
        raise AdapterError(
            ArtifactKind.TABLE,
            "table body must specify either `markdown` or `csv`",
        )

    if markdown is not None:
        html = _render_markdown(markdown)
    else:
        assert csv_src is not None  # for type narrowing
        html = _render_csv(csv_src)

    return RenderedArtifact(
        html_fragment=html,
        theme_class=theme_class,
        render_kind=ArtifactKind.TABLE,
        content_hash="",
        warnings=(),
    )


def _render_markdown(source: str) -> str:
    """Walk the markdown-it token stream and emit a single ``<table>``.

    Non-table events (paragraphs around the table) are ignored. If no
    table is present, raise :class:`AdapterError` so the producer learns
    about it (matches Rust ``table_markdown_without_table_rejected``).
    """

    md = MarkdownIt("commonmark").enable("table")
    tokens = md.parse(source)

    out: list[str] = []
    in_thead = False
    saw_table = False

    for tok in tokens:
        ttype = tok.type
        if ttype == "table_open":
            saw_table = True
            out.append(f'<table class="{WRAPPER_CLASS}">')
        elif ttype == "table_close":
            out.append("</table>")
        elif ttype == "thead_open":
            in_thead = True
            out.append("<thead>")
        elif ttype == "thead_close":
            in_thead = False
            out.append("</thead>")
        elif ttype == "tbody_open":
            out.append("<tbody>")
        elif ttype == "tbody_close":
            out.append("</tbody>")
        elif ttype == "tr_open":
            out.append(f'<tr class="{ROW_CLASS}">')
        elif ttype == "tr_close":
            out.append("</tr>")
        elif ttype == "th_open":
            out.append(f'<th class="{CELL_CLASS}">')
        elif ttype == "th_close":
            out.append("</th>")
        elif ttype == "td_open":
            out.append(f'<td class="{CELL_CLASS}">')
        elif ttype == "td_close":
            out.append("</td>")
        elif ttype == "inline":
            _emit_inline(out, tok.children or [], in_thead)
        # All other token types (paragraph_open/close, heading, etc.)
        # are dropped — table-only output.

    if not saw_table:
        raise AdapterError(
            ArtifactKind.TABLE, "markdown source contained no GFM table"
        )
    return "".join(out)


def _emit_inline(out: list[str], children: list, _in_thead: bool) -> None:
    """Emit inline children inside a cell.

    Mirrors the Rust adapter: ``text`` nodes are HTML-escaped; ``code``
    spans are wrapped in ``<code>``; any raw HTML is escaped (never
    passed through). Soft / hard breaks and other inline nodes are
    dropped — out of scope for table cells.
    """

    for child in children:
        ctype = child.type
        if ctype == "text":
            _push_escaped(out, child.content)
        elif ctype == "code_inline":
            out.append("<code>")
            _push_escaped(out, child.content)
            out.append("</code>")
        elif ctype in ("html_inline", "html_block"):
            # Never pass raw HTML through a cell — escape it as text so
            # a <script> in the producer's source surfaces as escaped
            # text, not executable markup.
            _push_escaped(out, child.content)
        # Drop everything else (softbreak, hardbreak, emphasis, etc.).


def _render_csv(source: str) -> str:
    """CSV → ``<table>``. First record is the header.

    Quoted fields with embedded commas (``"New York, NY"``) survive as
    one cell — that's stdlib ``csv``'s job.
    """

    reader = _csv.reader(io.StringIO(source))
    try:
        rows = list(reader)
    except _csv.Error as exc:
        raise AdapterError(
            ArtifactKind.TABLE, f"CSV parse failed: {exc}"
        ) from exc

    out: list[str] = []
    out.append(f'<table class="{WRAPPER_CLASS}">')

    if rows:
        headers = rows[0]
        if headers:
            out.append(f'<thead><tr class="{ROW_CLASS}">')
            for cell in headers:
                out.append(f'<th class="{CELL_CLASS}">')
                _push_escaped(out, cell)
                out.append("</th>")
            out.append("</tr></thead>")

        out.append("<tbody>")
        for record in rows[1:]:
            out.append(f'<tr class="{ROW_CLASS}">')
            for cell in record:
                out.append(f'<td class="{CELL_CLASS}">')
                _push_escaped(out, cell)
                out.append("</td>")
            out.append("</tr>")
        out.append("</tbody>")
    else:
        out.append("<tbody></tbody>")
    out.append("</table>")
    return "".join(out)


__all__ = ["CELL_CLASS", "ROW_CLASS", "WRAPPER_CLASS", "render"]
